kdd-2024/ibis-tutorial-KDD2024.md

```diff
@@ -0,0 +1,273 @@
+why Ibis, what is Ibis and for Who, I think this will let us introduce Ibis as part 
+of a "composable data ecosystem" (also requested by Ian).  We need to include 
+some streaming content and some ML content, again keep in mind that the ask is 
+to introduce this as a 20% of total content.
+
+## 0:00 - Intro and Setup “Going beyond pandas”
+
+Get attendees up and running in a GitHub Codespace or on their laptops. A bit of
+motivation about the kinds of problems where Ibis can help, and a general survey 
+of attendees to find out what their existing pain points and experiences are.
+
+## 0:15 - Introduction to Ibis basics
+
+**Slides**
+- What is Ibis?: Ibis as part of a composable data ecosystem.
+- Why Ibis?: present the problems that Ibis tackles and the benefits of adopting it.
+- Who Is Ibis for?: shown the versatility of profiles that can benefit from Ibis (data engineers, data analysts, and data scientists, any data practitioner) and how.
+
+**Get familiarized with basics: **
+
+A hands-on, follow-along notebook introducing the basic verbs of Ibis data 
+analysis (select, filter, group_by, order_by, and aggregate), demonstrating a 
+simple example on a few local query engines with hands-on exercises throughout.
+
+## 1:00 - Coffee Break (5 minutes that definitely takes 10 minutes)
+
+## 1:10 - In-memory tables, joins, and data analysis (need to adapt this for audience)
+
+Building on the previous notebook, we'll explore how to join in-memory data (from 
+a pandas DataFrame, Python dictionary, or PyArrow Table) with existing tables in 
+a local database and continue analysis on the join result.
+
+We'll touch on the Ibis deferred operator for specifying predicates in chained joins, 
+and demonstrate read_parquet and other read_* methods for loading local data into
+existing databases. Then we'll continue with a series of hands-on exercises, 
+building up an analysis pipeline for some IMDB ratings data, but only operating 
+on a 5% sample of the original dataset. After, we show how the same expression 
+can be computed on the full dataset without any code changes, both for local 
+execution, or with bursting to a cloud database (or other hosted database).
+
+## 1:50 - Coffee Break (5 minutes) + Q&A in the room
+
+HERE IS WHERE WE NEED TO THINK HOW TO TIE THE PREVIOUS CONTENT WITH STREAMING AND ML 
+WHAT WE HAVE IN THE SCIPY TUTORIAL IS 
+
+## 2:10 - Selectors 
+Continuing on from joins, we'll introduce selectors as a means of quickly 
+renaming and cleaning datasets, a powerful feature stolen inspired by dplyr.
+
+## 2:30 - UDFs and sql passthrough (I THINK THIS SHOULD BE GONE GIVEN THE AUDIENCE AND TIME)
+
+Demonstrate using UDFs to add custom operations.
+
+Explain and demonstrate various "escape hatches" when you really need to use SQL directly.
+
+# 3:00 - Coffee Break (5 minutes) + Q&A in the room
+
+## HELP HERE MAYBE AN EXAMPLE THAT CAN SHOW STREAMING AND ML, OR TWO? IT'LL HAVE TO BE SMALL
```

Comment:
+1 on things being disconnected. 

I think [something like this](https://ibis-project.org/posts/unified-stream-batch/) is a good example; generally the demos that we do for ibis streaming show some type of window aggregation in Flink (because window aggregations play an important role, especially in a lot of the ML-powered use cases that we were looking at), then execute the same window aggregation in DuckDB/another batch engine. 

My concern would be: would we need to spend time introducing Flink/streaming? I think it may be hard for the audience to understand why we're doing this / why this is important if they have little understanding of the problem space (the message may be blurry - they're all just backends, now a streaming backend, so what?). 

Perhaps 5 minutes talking about the problem / streaming and ~10 minutes walking through an example?

Reply:
@chloeh13q @deepyaman do either of you have any experience adding data generators like those used in the flink examples?

If it isn't too cumbersome, my thinking is we can have it fake incoming IMDB ratings, so we start with a batched version of the problem, using that to introduce Ibis as we build up expressions, then we can pivot to an example that involves a stream of incoming reviews (obviously contrived, but I think it would be reasonably representative)

Reply:
Yes we have done this before. Flink has a datagen connector; alternatively in the demos we spin up a Kafka topic and use the python-kafka client to write data into it